packages/transformer/src/IModelExporter.ts

```diff
@@ -268,9 +268,11 @@ export class IModelExporter {
    * @param user The user
    * @param startChangesetId Include changes from this changeset up through and including the current changeset.
    * If this parameter is not provided, then just the current changeset will be exported.
+   * @param changedInstanceIds Instance class that contains modified elements between 2 versions of an iModel.
+   * If this parameter is not provided, then [[ChangedInstanceIds.initialize]] will be called to discover changed elements.
    * @note To form a range of versions to export, set `startChangesetId` for the start (inclusive) of the desired range and open the source iModel as of the end (inclusive) of the desired range.
    */
-  public async exportChanges(user?: AccessToken, startChangesetId?: string): Promise<void> {
+  public async exportChanges(user?: AccessToken, startChangesetId?: string, changedInstanceIds?: ChangedInstanceIds): Promise<void> {
```

Comment:
instead of adding a third plain argument, can it be an object containing this, so that future additional arguments can be added in a non-breaking way? Eventually I intend to deprecate the overload passing a string argument for `startChangesetId` and instead also put that in the object

Reply:
It's not essential but yes, since it still introduces the new argument

Reply:
We should start doing this as a standard practice

Reply:
Yeah I'm willing to always bump minor on API changes, including non-breaking ones. It is better form.

Reply:
Also converted processChanges to use object as argument

Reply:
what you have now is a breaking change. I meant the following:

```ts
 public async exportChanges(user?: AccessToken, startChangesetId?: string, options: ExportChangesArgs): Promise<void> {
```

And then later we can deprecate that overload and do this. Although you can do so in this PR too, and we might as well if we're making it a minor release.

```ts
 /** @deprecated Don't use this overload */
 public async exportChanges(user?: AccessToken, startChangesetId?: string, options: ExportChangesArgs): Promise<void>
 /** desc */
 public async exportChanges(options: ExportChangesArgs): Promise<void>;
 public async exportChanges(userOrOptions?: AccessToken | ExportChangesArgs, startChangesetId?: string, options?: ExportChangesArgs): Promise<void> {
   const options = typeof userOrOptions === "string"
     ? {
       user: userOrOptions,
       startChangesetId,
       ...options
     }
     : userOrOptions; 
   //...
```

Reply:
Great idea!

Reply:
could you actually go ahead and add the deprecated overload as I showed above?

Reply:
Added

Reply:
do whatever works for VSCode, e.g. hovering later usages of `exportChanges` should show the description, but the invalid overload should be detected by both typescript (showing a ~strikethrough~ in vscode ) and eslint.

I believe it's really what I posted above. The full description remains on the valid overload signature which is declared first, then the deprecated overload is declared second with no description except `@deprecated in 0.1.x, use a single ExportChangesArgs object instead`, and that should work. Unless I removed them, I believe I've done this in other places.

Reply:
I changes descriptions. I was using itwinjs core lib as example.

Reply:
do you think you could cherry-pick in my version where I did something similar? 2f1189eafa68c7382c7e220030961157f487599a
You might also need c2f2da9d400f34e7b8b1c6b2fcd5d07fdae537a4...

Reply:
Is these changes still not in master? Maybe it would be better just merge this PR or close it and start over again

Reply:
There is multiple issues with those commits:
 - using types like `ChangesetIndexOrId` makes hardly readable code
 - importing types from different file, that can be reused, but actually is for different method is also makes code look like spaghetti

I would rather stick with my current version

Reply:
well the types are highly related, those functions call each other so they should have a relatively consistent API. And I disagree that `ChangesetIndexOrId` makes for hardly readable code. The value of using `ChangesetIndexOrId` is that the user can pass what they have without having to resolve the other kind using a query. This change actually reduced the amount of querying the transformer does which needs an index but was requesting an id.

If you're referring to the new "args" declarations as spaghetti, they are slightly unfortunate but that's a natural reality of supporting multiple overloads and makes sense until the next major version when we can drop the deprecated overload.

Are you still against this?

Reply:
I would rather have multiple types or interfaces which extends basic ones. Because there is a possibility to have a need to add additional property that will be used in one function and not in other function

Reply:
maybe I forgot a commit, there should be a second type that extends from the base.

Reply:
I think I forgot [ff19036f](https://github.com/iTwin/imodel-transformer/commit/ff19036ec6456c59891f6d56bdb2982e082d8d2f)
It's really a subset of federation-guid-optimization branch, which you could extract using
`git add -p federation-guid-optimization packages/transformer/src/IModelTransformer.ts packages/transformer/src/IModelExport.ts` and just ignore all other changes. Or try to cherry-pick those 3 commits but I didn't thoroughly check if that contains all the changes.

Reply:
I created #55 to demonstrate what I'd prefer